targets/unity-v2/source/Shared/Public/Pipeline/EventSendingStage.cs

```diff
@@ -0,0 +1,112 @@
+﻿#if NET_4_6
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using PlayFab.EventsModels;
+using PlayFab.Logger;
+using PlayFab.SharedModels;
+
+namespace PlayFab.Pipeline
+{
+    /// <summary>
+    /// The event sending stage.
+    /// </summary>
+    internal class EventSendingStage : PipelineStageBase<TitleEventBatch, PlayFabResult<WriteEventsResponse>>
+    {
+        private OneDSEventsAPI oneDSEventsApi;
+        private ILogger logger;
+
+        public EventSendingStage(ILogger logger)
+        {
+            this.logger = logger;
+            this.oneDSEventsApi = new OneDSEventsAPI();
+        }
+
+        /// <summary>
+        /// This method is called by pipeline for each available input item (a batch of events)
+        /// </summary>
+        /// <param name="batch">The input item (a batch of events).</param>
+        protected override void OnNextInputItem(TitleEventBatch batch)
+        {
+            // Send a batch and wait for result
+            var writeEventsRequest = new WriteEventsRequest
+            {
+                Events = batch.Events.Cast<PlayFabEmitEventRequest>().Select(x => x.Event.EventContents).ToList()
+            };
+
+            if (!oneDSEventsApi.IsOneDSAuthenticated)
+            {
+                var authResult = OneDSEventsAPI.GetTelemetryIngestionConfigAsync(new TelemetryIngestionConfigRequest());
+                try
+                {
+                    authResult.Wait(cts.Token);
+                    var response = authResult.Result.Result;
+                    if (response != null)
+                    {
+                        oneDSEventsApi.SetCredentials("o:" + response.TenantId, response.IngestionKey, response.TelemetryJwtToken, response.TelemetryJwtHeaderKey, response.TelemetryJwtHeaderPrefix);
+                    }
+                    else
+                    {
+                        throw new Exception($"Failed to get OneDS authentication token from PlayFab service");
+                    }
+                }
+                catch (Exception e)
+                {
+                    // Cancel result promises that will never be fulfilled
+                    // and move on to the next batch. 
+                    foreach (var request in batch.Events)
+                    {
+                        PlayFabEmitEventRequest eventRequest = (PlayFabEmitEventRequest)request;
+                        eventRequest.ResultPromise?.SetCanceled();
+                    }
+
+                    logger.Error($"Exception in OnNextInputItem {e.Source} with message: {e}.");
+                }
+            }
+
+            Task<PlayFabResult<WriteEventsResponse>> apiTask = oneDSEventsApi.WriteTelemetryEventsAsync(writeEventsRequest);
+            try
+            {
+                apiTask.Wait(cts.Token);
+                FulfillPromises(batch.Events, apiTask.Result);
+            }
+            catch (Exception e)
```

Comment:
This is better, because it's just two lines, so I have a pretty good idea where the exception will occur.